src/Lykke.AlgoStore.Api/RealTimeStreaming/Sources/RabbitMq/ObservableRabbitMqConnection.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reactive.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Common.Log;
+using Lykke.AlgoStore.Api.RealTimeStreaming.DataTypes;
+using Lykke.Common.Log;
+using Lykke.RabbitMqBroker;
+using Lykke.RabbitMqBroker.Subscriber;
+
+namespace Lykke.AlgoStore.Api.RealTimeStreaming.Sources.RabbitMq
+{
+    public class ObservableRabbitMqConnection<T> : RealTimeDataSourceBase<T> where T : BaseDataModel, new()
+    {
+        private readonly IObservable<T> _messages;
+        private RabbitMqSubscriber<T> _subscriber;
+        private readonly ILogFactory _logFactory; //https://github.com/LykkeCity/Lykke.Logs/blob/master/migration-to-v5.md
+        private readonly BlockingCollection<T> _messageQueue;
+        private readonly RabbitMqSubscriptionSettings _rabbitSettings;
+
+        public ObservableRabbitMqConnection(RabbitMqSubscriptionSettings rabbitSettings, ILogFactory logFactory)
+        {
+            _messages = Observable.Create<T>(async (obs) => { await ReadRabbitMqMessagesLoop(obs); });
+            _messageQueue = new BlockingCollection<T>();
+            _rabbitSettings = rabbitSettings;
+            _logFactory = logFactory;
+            InitializeRabbitMqConnection();
+        }
+
+        private void InitializeRabbitMqConnection()
+        {
+            _subscriber = new RabbitMqSubscriber<T>(_logFactory, _rabbitSettings, new DefaultErrorHandlingStrategy(_logFactory, _rabbitSettings))
+                .SetMessageDeserializer(new GenericRabbitModelConverter<T>())
+                .SetMessageReadStrategy(new MessageReadWithTemporaryQueueStrategy())
+                .CreateDefaultBinding()
+                .Subscribe(OnMessageReceived);
+        }
+
+        private Task OnMessageReceived(T message)
+        {
+            if (message.IsAllowedByFilter(Filter))
+            {
+                _messageQueue.TryAdd(message, TimeSpan.FromSeconds(5));
+            }
+            return Task.CompletedTask;
+        }
+
+        public override IDisposable Subscribe(IObserver<T> observer)
+        {
+            return _messages.Subscribe(observer);
+        }
+
+        private async Task ReadRabbitMqMessagesLoop(IObserver<T> obs)
+        {
+            await Task.Run(() =>
```

Comment:
You can avoid using Task.Run by reworking the while loop - instead of using `_messageQueue.TryTake` you can replace the BlockingCollection with a ConcurrentQueue and have the if statement go something like this

```c#
if(_messageQueue.TryDequeue(out T message))
    obs.OnNext(message);
else
    await Task.Delay(TimeSpan.FromSeconds(2));
```